### DISCUSSION TITLE
Why does the lockfile contains dozens of versions of each package?

### DISCUSSION THREAD

**Discussion by BlueskyFR** [python-poetry/poetry#7790] [Q&A]
The lockfile has this for numpy for instance, why is that?
![image](https://user-images.githubusercontent.com/25421460/231478965-ec37700c-392e-43aa-8cbb-5df92e73ef9b.png)

**Comment by radoering** [comment]
As far as I can see in the screenshot, there is just one version (namely 1.24.2) in the lockfile. There all all distributions of this version in the lockfile because poetry's lockfile is environment independent. (If different versions are required for different environments, there will be different versions in the lockfile.)

**Reply by BlueskyFR** [reply]
oh ok so you mean Poetry does not actually download all of them and just lists them?

**Reply by radoering** [reply]
IIRC, it depends on the source. If the source provides hashes (like PyPI for example), Poetry will not download all of them. If the source does not provide hashes, Poetry will probably download all of them to calculate the hashes.
